### PR TITLE
Add type hints to main APIs

### PR DESCRIPTION
**Pull request by joseph-pq**
# Related Issues
#263 
# Changes
Added typehints to files:
- edfreader.py
- edfwriter.py
- highleve.py

Also, I added a `py.typed` file for being recognizable for mypy and a `_pyedflib.pyi` as a python interface for the _pyedflib.pyc.

### CODE REVIEW

File: pyedflib/highlevel.py [hunk 1]
```diff
@@ -217,9 +233,18 @@ def make_header(technician='', recording_additional='', patientname='',
...
 
 
-def make_signal_header(label, dimension='uV', sample_rate=256, sample_frequency=None,
-                       physical_min=-200, physical_max=200, digital_min=-32768,
-                       digital_max=32767, transducer='', prefiler=''):
+def make_signal_header(
+    label: str,
+    dimension: str = 'uV',
+    sample_rate: int = 256,
+    sample_frequency: Optional[int] = None,
```
Comment (skjerns) [inline pyedflib/highlevel.py:240]:
```suggestion
    sample_rate: Union[int, float] = 256,
    sample_frequency: Optional[Union[int, float]] = None,
```

File: pyedflib/highlevel.py [hunk 2]
```diff
@@ -217,9 +233,18 @@ def make_header(technician='', recording_additional='', patientname='',
...
+    physical_min: float = -200,
+    physical_max: float = 200,
+    digital_min: int = -32768,
+    digital_max: int = 32767,
```
Comment (skjerns) [inline pyedflib/highlevel.py:244]:
```suggestion
    digital_min: Union[float, int] = -32768,
    digital_max:  Union[float, int] = 32767,
```

File: pyedflib/highlevel.py [hunk 3]
```diff
@@ -269,10 +294,18 @@ def make_signal_header(label, dimension='uV', sample_rate=256, sample_frequency=
...
 
-def make_signal_headers(list_of_labels, dimension='uV', sample_rate=256,
-                       sample_frequency=None, physical_min=-200.0, physical_max=200.0,
-                       digital_min=-32768, digital_max=32767,
-                       transducer='', prefiler=''):
+def make_signal_headers(
+    list_of_labels: List[str],
+    dimension: str = 'uV',
+    sample_rate: int = 256,
+    sample_frequency: Optional[int] = None,
```
Comment (skjerns) [inline pyedflib/highlevel.py:301]:
```suggestion
    sample_frequency: Optional[Union[int, float]] = None,
```

File: pyedflib/highlevel.py [hunk 4]
```diff
@@ -269,10 +294,18 @@ def make_signal_header(label, dimension='uV', sample_rate=256, sample_frequency=
...
+    physical_min: float = -200.0,
+    physical_max: float = 200.0,
+    digital_min: int = -32768,
+    digital_max: int = 32767,
```
Comment (skjerns) [inline pyedflib/highlevel.py:305]:
```suggestion
    digital_min: Union[float,int] = -32768,
    digital_max: Union[float, int] = 32767,
```

File: pyedflib/highlevel.py [hunk 5]
```diff
@@ -507,7 +552,12 @@ def write_edf(edf_file, signals, signal_headers, header=None, digital=False,
     return os.path.isfile(edf_file)
 
 
-def write_edf_quick(edf_file, signals, sfreq, digital=False):
+def write_edf_quick(
+    edf_file: str,
+    signals: np.ndarray,
+    sfreq: int,
```
Comment (skjerns) [inline pyedflib/highlevel.py:558]:
```suggestion
    sfreq: Union[float, int],
```

File: pyedflib/edfwriter.py [hunk 6]
```diff
@@ -978,7 +984,7 @@ def _calculate_optimal_record_duration(self):
         assert record_duration<=60, 'record duration must be below 60 seconds'
         self.record_duration = record_duration
 
-    def _get_sample_frequency(self, channelIndex):
+    def _get_sample_frequency(self, channelIndex: int) -> Optional[Union[int, float]]:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:987]:
```suggestion
    def _get_sample_frequency(self, channelIndex: int) -> Union[int, float]:
```

should always return something

File: pyedflib/edfwriter.py [hunk 7]
```diff
@@ -136,29 +142,28 @@ def check_signal_header_correct(channels, i, file_type):
...
+def u(x: bytes) -> str:
     return x.decode("utf_8", "strict")
 
 
-def du(x):
+def du(x: Union[str, bytes]) -> bytes:
     if isinstance(x, bytes):
         return x
     else:
         return x.encode("utf_8")
```
Comment (skjerns) [inline pyedflib/edfwriter.py:153]: I'm actually not sure these functions are necessary after deprecating Python 2.7.

File: pyedflib/edfwriter.py [hunk 8]
```diff
@@ -136,29 +142,28 @@ def check_signal_header_correct(channels, i, file_type):
...
 
 
-def isbytestr(s):
+def isbytestr(s: Any) -> bool:
     warnings.warn("Function 'isbytestr' is deprecated.", DeprecationWarning, stacklevel=2)
     return isinstance(s, bytes)
 
 
-def sex2int(sex):
+def sex2int(sex: Union[int, str, None]) -> Optional[int]:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:166]: Technically it always returns an `int` or `None`, but can raise an Exception, would that still be `Optional` then?

```suggestion
def sex2int(sex: Union[int, str, None]) -> Optional[int, None]:
```

File: pyedflib/edfwriter.py [hunk 9]
```diff
@@ -171,39 +176,38 @@ def sex2int(sex):
         raise ValueError(f"Unknown sex: '{sex}'")
 
 
-def gender2int(gender):
+def gender2int(gender: Union[int, str, None]) -> Optional[int]:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:179]: same as above

File: pyedflib/edfwriter.py [hunk 10]
```diff
@@ -171,39 +176,38 @@ def sex2int(sex):
...
         self.parameter = value
 
-    def __str__(self):
+    def __str__(self) -> str:
         return repr(self.parameter)
 
 
 class EdfWriter:
-    def __exit__(self, exc_type, exc_val, ex_tb):
+    def __exit__(self, exc_type: Optional[type], exc_val: Optional[BaseException], ex_tb: Optional[BaseException]) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:201]:
```suggestion
    def __exit__(self, exc_type: Optional[type], exc_val: Optional[BaseException], ex_tb: Optional[BaseException]) -> None:
    def __exit__(self, exc_type: Optional[Type[BaseException]], exc_val: Optional[BaseException], exc_tb: Optional[TracebackType]) -> None:
)
    
```

according to https://stackoverflow.com/questions/58808055/what-are-the-python-builtin-exit-argument-types

would also need to import `from types import TracebackType`

File: pyedflib/edfwriter.py [hunk 11]
```diff
@@ -336,7 +340,7 @@ def setHeader(self, fileHeader):
         self.birthdate = fileHeader["birthdate"]
         self.update_header()
 
-    def setSignalHeader(self, edfsignal, channel_info):
+    def setSignalHeader(self, edfsignal: int, channel_info: Dict[str, Union[str, int, float]]) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:343]:
```suggestion
    def setSignalHeader(self, edfsignal: int, channel_info: Dict[str, Union[str, int, float, None]]) -> None:
```

File: pyedflib/edfwriter.py [hunk 12]
```diff
@@ -357,7 +361,7 @@ def setSignalHeader(self, edfsignal, channel_info):
         self.channels[edfsignal].update(channel_info)
         self.update_header()
 
-    def setSignalHeaders(self, signalHeaders):
+    def setSignalHeaders(self, signalHeaders: List[Dict[str, Union[str, int, float]]]) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:364]:
```suggestion
    def setSignalHeaders(self, signalHeaders: List[Dict[str, Union[str, int, float, None]]]) -> None:
```

File: pyedflib/edfwriter.py [hunk 13]
```diff
@@ -557,7 +561,7 @@ def setStartdatetime(self, recording_start_time):
         self.recording_start_time = recording_start_time
         self.update_header()
 
-    def setBirthdate(self, birthdate):
+    def setBirthdate(self, birthdate: date) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:564]:
```suggestion
    def setBirthdate(self, birthdate: Union[str, date]) -> None:
```

File: pyedflib/edfwriter.py [hunk 14]
```diff
@@ -606,7 +610,7 @@ def setSamplefrequency(self, edfsignal, samplefrequency):
         self.channels[edfsignal]['sample_frequency'] = samplefrequency
         self.update_header()
 
-    def setPhysicalMaximum(self, edfsignal, physical_maximum):
+    def setPhysicalMaximum(self, edfsignal: int, physical_maximum: float) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:613]:
```suggestion
    def setPhysicalMaximum(self, edfsignal: int, physical_maximum: Union[int, float]) -> None:
```

File: pyedflib/edfwriter.py [hunk 15]
```diff
@@ -626,7 +630,7 @@ def setPhysicalMaximum(self, edfsignal, physical_maximum):
         self.channels[edfsignal]['physical_max'] = physical_maximum
         self.update_header()
 
-    def setPhysicalMinimum(self, edfsignal, physical_minimum):
+    def setPhysicalMinimum(self, edfsignal: int, physical_minimum: float) -> None:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:633]:
```suggestion
    def setPhysicalMinimum(self, edfsignal: int, physical_minimum: Union[in, float]) -> None:
```

File: pyedflib/edfwriter.py [hunk 16]
```diff
@@ -904,7 +908,7 @@ def writeSamples(self, data_list, digital = False):
                 if success<0:
                     raise OSError(f'Unknown error while calling writeSamples: {success}')
 
-    def writeAnnotation(self, onset_in_seconds, duration_in_seconds, description, str_format='utf_8'):
+    def writeAnnotation(self, onset_in_seconds: float, duration_in_seconds: float, description: str, str_format: str = 'utf_8') -> int:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:911]:
```suggestion
    def writeAnnotation(self, onset_in_seconds: Union[int, float], duration_in_seconds: Union[int, float], description: str, str_format: str = 'utf_8') -> int:
```

File: pyedflib/edfwriter.py [hunk 17]
```diff
@@ -921,18 +925,20 @@ def writeAnnotation(self, onset_in_seconds, duration_in_seconds, description, st
...
-    def close(self):
+    def close(self) -> None:
         """
         Closes the file.
         """
         close_file(self.handle)
         self.handle = -1
 
-    def get_smp_per_record(self, ch_idx):
+    def get_smp_per_record(self, ch_idx: int) -> Optional[int]:
```
Comment (skjerns) [inline pyedflib/edfwriter.py:941]:
```suggestion
    def get_smp_per_record(self, ch_idx: int) -> int:
```

File: pyedflib/edfreader.py [hunk 18]
```diff
@@ -59,60 +68,84 @@ def _debug_parse_header(filename): #pragma: no cover
...
+
+    def __enter__(self) -> "EdfReader":
         return self
 
-    def __del__(self):
+    def __del__(self) -> None:
         self._close()
 
-    def __exit__(self, exc_type, exc_val, ex_tb):
+    def __exit__(self, exc_type, exc_val, ex_tb) -> None:
```
Comment (skjerns) [inline pyedflib/edfreader.py:112]:
```suggestion
def __exit__(
    self,
    exc_type: Optional[Type[BaseException]],
    exc_val: Optional[BaseException],
    exc_tb: Optional[TracebackType],
) -> None:
```

see comment below, need to import tracebacktype as well

File: pyedflib/edfreader.py [hunk 19]
```diff
@@ -124,60 +157,69 @@ def _get_float(self, v):
                 result[i] = float(v[i])
         return result
 
-    def _convert_string(self,s):
+    # FIX: Seems s must always be bytes because string has no decode method
+    def _convert_string(self, s: Union[bytes, str]) -> str:
```
Comment (skjerns) [inline pyedflib/edfreader.py:161]:
```suggestion
    def _convert_string(self, s: Union[bytes]) -> str:
```

how was the `str` suggested? Is the method called with `str` somewhere?

File: pyedflib/edfreader.py [hunk 20]
```diff
@@ -124,60 +157,69 @@ def _get_float(self, v):
...
         if isinstance(s, bytes):
             return s.decode("latin")
         else:
-            return s.decode("utf_8", "strict")
+            return s.decode("utf_8", "strict")  # type: ignore
```
Comment (skjerns) [inline pyedflib/edfreader.py:165]:
```suggestion
            return s  # type: ignore
```

Comment (skjerns) [inline pyedflib/edfreader.py:165]: could you change to `elif isinstance(s, str): return s`?

File: pyedflib/edfreader.py [hunk 21]
```diff
@@ -592,15 +664,18 @@ def getDigitalMaximum(self, chn=None):
...
+                    )
+                )
         else:
             digMax = np.zeros(self.signals_in_file)
             for i in np.arange(self.signals_in_file):
                 digMax[i] = self.digital_max(i)
             return digMax
 
-    def getDigitalMinimum(self, chn=None):
+    def getDigitalMinimum(self, chn: Optional[int] = None):
```
Comment (skjerns) [inline pyedflib/edfreader.py:678]:
```suggestion
    def getDigitalMinimum(self, chn: Optional[int] = None) -> Union[int, np.ndarray]:
```

File: pyedflib/edfreader.py [hunk 22]
```diff
@@ -674,10 +756,15 @@ def getPhysicalDimension(self, chn):
...
+            raise IndexError(
+                "Trying to access channel {}, but only {} " "channels found".format(
+                    chn, self.signals_in_file
+                )
+            )
 
-    def readSignal(self, chn, start=0, n=None, digital=False):
+    def readSignal(
+        self, chn: int, start: int = 0, n: Optional[int] = None, digital: bool = False
+    ):
```
Comment (skjerns) [inline pyedflib/edfreader.py:767]:
```suggestion
    def readSignal(
        self, chn: int, start: int = 0, n: Optional[int] = None, digital: bool = False
    ) -> np.ndarray:
```

File: pyedflib/_extensions/_pyedflib.pyi [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+def is_file_used(path: str) -> bool: ...
+def set_physical_maximum(handle: int, edfsignal: int, phys_max: float) -> int: ...
+def open_file_writeonly(path: str, filetype: int, number_of_signals: int) -> int: ...
+def set_patient_additional(handle: int, patient_additional: Union[str, bytes]) -> int: ...
+def set_digital_maximum(handle: int, edfsignal: int, dig_max: int) -> int: ...
+def set_birthdate(handle: int, birthdate_year: int, birthdate_month: int, birthdate_day: int) -> int: ...
+def set_digital_minimum(handle: int, edfsignal: int, dig_min: int) -> int: ...
+def write_digital_samples(handle: int, buf: np.ndarray[np.int32_t]) -> int: ...
+def set_equipment(handle: int, equipment: Union[str, bytes]) -> int: ...
+def set_samples_per_record(handle: int, edfsignal: int, smp_per_record: Optional[int]) -> int: ...
```
Comment (skjerns) [inline pyedflib/_extensions/_pyedflib.pyi:146]:
```suggestion
def set_samples_per_record(handle: int, edfsignal: int, smp_per_record: Optional[int, None]) -> int: ...
```

**Review by skjerns — COMMENTED** [review]
Thanks! looking good :) if you implement the changes, I will merge

File: pyedflib/edfwriter.py (hunk 8 above)
Comment (joseph-pq) [reply pyedflib/edfwriter.py:166]: As specified in [PEP 484](https://peps.python.org/pep-0484/#exceptions), exceptions are out of scope of type hints. I guess it won't be required to add None as inside the Optional neither.

File: pyedflib/edfreader.py (hunk 20 above)
Comment (joseph-pq) [reply pyedflib/edfreader.py:173]: Done

**Review by joseph-pq — COMMENTED** [review]
@skjerns I accepted most of the suggestions. I responded to the rest.

File: pyedflib/edfwriter.py (hunk 7 above)
Comment (skjerns) [reply pyedflib/edfwriter.py:154]: good idea! i created an issue: https://github.com/holgern/pyedflib/issues/265

**Comment by skjerns** [comment]
@joseph-pq there seems to be a syntax error, could you check?

**Comment by skjerns** [comment]
@holgern could you enable `squash and merge` as the default for the repo settings?

image

File: pyedflib/edfwriter.py (hunk 8 above)
Comment (skjerns) [reply pyedflib/edfwriter.py:167]: Yep indeed didn't know that! Thanks :)

File: pyedflib/edfwriter.py (hunk 9 above)
Comment (joseph-pq) [reply pyedflib/edfwriter.py:180]: I guess I didn't explain myself clearly. What I meant was that the `gender2int` function should have the same type hint as `sex2int`.

Comment (skjerns) [reply pyedflib/edfwriter.py:180]: Sure!

**Comment by skjerns** [comment]
thanks to @joseph-pq for the great PR, will merge

**Pull request merged**